### PR TITLE
Add multi-class support

### PR DESCRIPTION
**Pull request by Ignose**
Updated stat checks to be generic, reagent checks to work for all three classes, etc. 

Non-Mys classes REQUIRE Prevent Scurvy and Sobriety.

Sorry about the new ESLint and Libram versions.

### CODE REVIEW

File: KoLmafia/scripts/instantsccs/instantsccs.js
Comment (Pantocyclus) [inline KoLmafia/scripts/instantsccs/instantsccs.js:1]: Do not upload the compiled js file into the main branch of the repo. Ignore it with gitignore

File: package.json [hunk 1]
```diff
@@ -32,8 +32,8 @@
   "dependencies": {
     "core-js": "^3.16.4",
     "grimoire-kolmafia": "^0.3.18",
-    "kolmafia": "^5.27578.0",
-    "libram": "^0.8.9"
+    "kolmafia": "^5.27609.0",
```
Comment (Pantocyclus) [inline package.json:35]: If we are using an updated version of KoLMafia, then we should also require that users have at least that version (where we enforce this with `sinceKolmafiaRevision()` in `main.ts`)

File: src/combat.ts [hunk 2]
```diff
@@ -1,7 +1,8 @@
-import { mpCost, myClass, toInt } from "kolmafia";
-import { $item, $skill, get, have, StrictMacro } from "libram";
+import { mpCost, myClass, myPrimestat, toInt } from "kolmafia";
+import { $item, $skill, $stat, get, have, StrictMacro } from "libram";
 
-export const mainStat = myClass().primestat;
+//export const mainStat = myClass().primestat;
+export const mainStat = myPrimestat(); //Update to select mainstat based on class derived from Libram
```
Comment (Pantocyclus) [inline src/combat.ts:5]: The comments here are unnecessary. You can remove them

File: src/combat.ts [hunk 3]
```diff
@@ -11,17 +12,30 @@ export default class Macro extends StrictMacro {
...
+      //Update to add alternative combat options for non-Mys classes
+      return (useCinch ? macroHead.trySkill($skill`Cincho: Confetti Extravaganza`) : macroHead)
+        .while_(
+          `!mpbelow ${mpCost($skill`Saucegeyser`)} && hasskill ${toInt($skill`Saucegeyser`)}`,
+          Macro.skill($skill`Saucegeyser`)
+        )
+        .while_(
+          `!mpbelow ${mpCost($skill`Saucestorm`)} && hasskill ${toInt($skill`Saucestorm`)}`,
+          Macro.skill($skill`Saucestorm`)
+        )
```
Comment (Pantocyclus) [inline src/combat.ts:25]: This should be the default for all non-mus classes, we probably shouldn't be using Lunging Thrust-Smack for mox classes.

Also, this should be refactored into something like
```
const defaultClassAttack = mainStat === $stat`Muscle` ? Macro.while_(/*LTS logic*/) : Macro.while(/*spellslinging logic/*);
return (useCinch ? macroHead.trySkill($skill`Cincho: Confetti Extravaganza`) : macroHead).defaultClassAttack.attack().repeat()
```

File: src/combat.ts [hunk 4]
```diff
@@ -11,17 +12,30 @@ export default class Macro extends StrictMacro {
...
-        `!mpbelow ${mpCost($skill`Saucegeyser`)} && hasskill ${toInt($skill`Saucegeyser`)}`,
-        Macro.skill($skill`Saucegeyser`)
-      )
-      .while_(
-        `!mpbelow ${mpCost($skill`Saucestorm`)} && hasskill ${toInt($skill`Saucestorm`)}`,
-        Macro.skill($skill`Saucestorm`)
-      )
-      .attack()
-      .repeat();
+    if (myPrimestat() === $stat`Mysticality`) {
```
Comment (Pantocyclus) [inline src/combat.ts:15]: We defined `mainStat` above, so there's no reason to call `myPrimestat()` here. Alternatively, we should just use `myPrimestat()` wherever we need it and not need to define and export `mainStat` above

File: src/engine/outfit.ts [hunk 5]
```diff
@@ -93,9 +94,11 @@ function optimisticCandle(): Familiar {
 }
 
 function melodramedary(): Familiar {
-  return have($familiar`Melodramedary`) &&
+  return (have($familiar`Melodramedary`) &&
     camelFightsLeft() >= Math.ceil((100 - get("camelSpit")) / 3.0) &&
-    get("camelSpit") < 100
+    get("camelSpit") < 100) || (have($familiar`Melodramedary`) &&
+    camelFightsLeft() >= Math.ceil((100 - get("camelSpit")) / 4.0) &&
+    get("camelSpit") < 100 && have($item`dromedary drinking helmet`))
```
Comment (Pantocyclus) [inline src/engine/outfit.ts:101]: Both sides of the `A || B` have `get("camelSpit") < 100`, so it should be brought out of the or statement.

I also do not agree with this in principle. The division by 4.0 is an expectation, but you are not guaranteed to get spit by that number of turns - if it turns out that you're unlucky it could turn out to be extremely inefficient. I don't think there is any other logic in the script that forgoes a guaranteed strat for a probabilistic one

File: src/engine/outfit.ts [hunk 6]
```diff
@@ -141,16 +144,27 @@ export function baseOutfit(allowAttackingFamiliars = true): OutfitSpec {
...
+    weapon: have($item`fish hatchet`)
+      ? $item`fish hatchet`
+      : have($item`bass clarinet`)
+      ? $item`bass clarinet`
+      : undefined,
     hat: avoidDaylightShavingsHelm() ? undefined : $item`Daylight Shavings Helmet`,
     offhand: $item`unbreakable umbrella`,
     back: lovTunnelCompleted ? $item`LOV Epaulettes` : undefined,
-    acc1: $item`codpiece`,
+    acc1: have($item`codpiece`) ? $item`codpiece` : undefined,
```
Comment (Pantocyclus) [inline src/engine/outfit.ts:155]: Why is this necessary? Is this not already properly handled in `engine.ts`?

File: src/engine/outfit.ts [hunk 7]
```diff
@@ -141,16 +144,27 @@ export function baseOutfit(allowAttackingFamiliars = true): OutfitSpec {
...
     acc2:
       have($item`Cincho de Mayo`) && get("_cinchUsed", 0) < 95 && !get("instant_saveCinch", false)
         ? $item`Cincho de Mayo`
         : undefined,
     familiar: chooseFamiliar(allowAttackingFamiliars),
-    modifier: "0.25 mys, 0.33 ML, -equip tinsel tights, -equip wad of used tape",
+    famequip:
+      have($item`dromedary drinking helmet`) && chooseFamiliar() == $familiar`melodramedary`
```
Comment (Pantocyclus) [inline src/engine/outfit.ts:162]: `chooseFamiliar(allowAttackingFamiliars)`

File: src/tasks/leveling.ts [hunk 8]
```diff
@@ -392,46 +450,108 @@ export const LevelingQuest: Quest = {
...
       // TODO: Make this completed if we've already wished twice with the paw (requires mafia tracking)
       completed: () =>
-        have($effect`Different Way of Seeing Things`) ||
+        have(xpWishEffect) ||
         !have($item`cursed monkey's paw`) ||
-        forbiddenEffects.includes($effect`Different Way of Seeing Things`) ||
+        forbiddenEffects.includes(xpWishEffect) ||
         get("instant_saveMonkeysPaw", false) ||
-        myBasestat($stat`Mysticality`) >= targetBaseMyst - targetBaseMystGap ||
+        myBasestat(myPrimestat()) >= targetBaseMyst - targetBaseMystGap ||
```
Comment (Pantocyclus) [inline src/tasks/leveling.ts:457]: We should rename `targetBaseMyst` to `targetBasestat`, and also properly handle the transition of the relevant preference

File: src/tasks/leveling.ts [hunk 9]
```diff
@@ -392,46 +450,108 @@ export const LevelingQuest: Quest = {
...
+          .split(",")
+          .includes(toInt($item`tobiko marble soda`).toString()) ||
+        have($item`tobiko marble soda`) ||
+        storageAmount($item`tobiko marble soda`) === 0 ||
+        !get("instant_experimentPulls", true),
       do: (): void => {
-        takeStorage($item`abstraction: category`, 1);
-        chew($item`abstraction: category`, 1);
+        takeStorage($item`tobiko marble soda`, 1);
       },
```
Comment (Pantocyclus) [inline src/tasks/leveling.ts:555]: I don't think these experimental pulls should be added to main in this manner or at this stage

File: src/tasks/leveling.ts [hunk 10]
```diff
@@ -754,33 +878,37 @@ export const LevelingQuest: Quest = {
...
+        );
+        if (!have(reagentBoosterItem)) {
           if (get("reagentSummons") === 0) useSkill($skill`Advanced Saucecrafting`, 1);
-          create($item`ointment of the occult`, 1);
+          create(reagentBoosterItem, 1);
         }
-        ensureEffect($effect`Mystically Oiled`);
+        ensureEffect(reagentBoosterEffect);
       },
     },
```
Comment (Pantocyclus) [inline src/tasks/leveling.ts:895]: How are we guaranteeing an olive for mox classes at this point?

File: src/tasks/leveling.ts [hunk 11]
```diff
@@ -938,7 +1066,7 @@ export const LevelingQuest: Quest = {
     {
       name: "Monster Habitats",
       ready: () =>
-        get("monsterHabitatsFightsLeft") > 0 &&
+        toInt(get("monsterHabitatsFightsLeft")) > 0 &&
```
Comment (Pantocyclus) [inline src/tasks/leveling.ts:1069]: Why is this necessary? Libram should already be casting it to the right datatype

File: src/tasks/leveling.ts [hunk 12]
```diff
@@ -948,13 +1076,14 @@ export const LevelingQuest: Quest = {
         usefulEffects.forEach((ef) => tryAcquiringEffect(ef));
         restoreMp(50);
       },
-      completed: () => get("monsterHabitatsFightsLeft") === 0,
+      completed: () => toInt(get("monsterHabitatsFightsLeft")) === 0,
```
Comment (Pantocyclus) [inline src/tasks/leveling.ts:1079]: Same comment as above

File: src/tasks/leveling.ts [hunk 13]
```diff
@@ -1153,8 +1282,8 @@ export const LevelingQuest: Quest = {
       do: () => Witchess.fightPiece($monster`Witchess Bishop`),
       combat: new CombatStrategy().macro(() =>
         Macro.externalIf(
-          get("monsterHabitatsFightsLeft") <= 1 &&
-            get("_monsterHabitatsRecalled") < 3 - get("instant_saveMonsterHabitats", 0) &&
+          toInt(get("monsterHabitatsFightsLeft")) <= 1 &&
+            toInt(get("_monsterHabitatsRecalled")) < 3 - get("instant_saveMonsterHabitats", 0) &&
```
Comment (Pantocyclus) [inline src/tasks/leveling.ts:1286]: Ditto

File: src/tasks/leveling.ts [hunk 14]
```diff
@@ -1192,7 +1321,7 @@ export const LevelingQuest: Quest = {
     {
       name: "Powerlevel",
       completed: () =>
-        myBasestat($stat`Mysticality`) >= targetBaseMyst - targetBaseMystGap &&
+        myBasestat(myPrimestat()) >= targetBaseMyst - targetBaseMystGap && // I don't know if this will cause issues.
```
Comment (Pantocyclus) [inline src/tasks/leveling.ts:1324]: This is fine, we can test it and update it as necessary, but leave the comments out of the code

File: src/tasks/leveling.ts [hunk 15]
```diff
@@ -1412,7 +1541,7 @@ export const LevelingQuest: Quest = {
       do: () => CombatLoversLocket.reminisce($monster`Witchess King`),
       combat: new CombatStrategy().macro(() =>
         Macro.externalIf(
-          get("monsterHabitatsFightsLeft") <= 1 &&
+          toInt(get("monsterHabitatsFightsLeft")) <= 1 &&
```
Comment (Pantocyclus) [inline src/tasks/leveling.ts:1544]: Ditto to `toInt()`

File: src/tasks/leveling.ts [hunk 16]
```diff
@@ -1454,7 +1583,7 @@ export const LevelingQuest: Quest = {
           };
       },
       completed: () =>
-        myBasestat($stat`Mysticality`) >= targetBaseMyst &&
+        myBasestat(myPrimestat()) >= targetBaseMyst &&
```
Comment (Pantocyclus) [inline src/tasks/leveling.ts:1586]: Ditto for `targetBaseMyst`

File: src/tasks/runstart.ts [hunk 17]
```diff
@@ -123,13 +132,21 @@ export const RunStartQuest: Quest = {
...
+        } else if (myPrimestat() === $stat`Mysticality`) {
+          retrieveItem($item`codpiece`, 1);
+          use($item`codpiece`, 1);
+          create($item`oil cap`, 1);
+          autosell($item`oil cap`, 1);
+        } else if (myPrimestat() === $stat`Moxie`) {
+          retrieveItem($item`bass clarinet`);
+          use($item`bass clarinet`, 1);
+          autosell($item`white pixel`, 10);
+        }
```
Comment (Pantocyclus) [inline src/tasks/runstart.ts:149]: The lack of meat for muscle classes might

Comment (Pantocyclus) [inline src/tasks/runstart.ts:149]: How will this potentially affect meat sources for low shinies? I'm also assuming there are no significant conflict of NC% slots for the NC test

File: src/tasks/runstart.ts [hunk 18]
```diff
@@ -226,6 +243,21 @@ export const RunStartQuest: Quest = {
...
+        !have($familiar`Melodramedary`) ||
+        !get("instant_camelExperiment", true),
+      do: (): void => {
+        if (!have($item`box of Familiar Jacks`)) create($item`box of Familiar Jacks`, 1);
+
+        useFamiliar($familiar`Melodramedary`);
+        use($item`box of Familiar Jacks`, 1);
+      },
+      limit: { tries: 1 },
+    },
```
Comment (Pantocyclus) [inline src/tasks/runstart.ts:260]: This experimental task should be left out for now

File: src/tasks/runstart.ts [hunk 19]
```diff
@@ -245,7 +277,8 @@ export const RunStartQuest: Quest = {
       completed: () => !have($item`sugar sheet`),
       do: (): void => {
         const nextMissingSugarItem =
-          $items`sugar shorts, sugar chapeau, sugar shank`.find((it) => !have(it)) || $item`none`;
+          $items`sugar shorts, sugar chapeau, sugar shank, sugar shield`.find((it) => !have(it)) ||
```
Comment (Pantocyclus) [inline src/tasks/runstart.ts:280]: Addressed the sugar shield query above

File: src/tasks/runstart.ts [hunk 20]
```diff
@@ -331,7 +367,20 @@ export const RunStartQuest: Quest = {
...
+          visitUrl(
+            "choice.php?pwd&option=1&whichchoice=1331&g=" +
+              (random(2) + 1) +
+              "&local[]=2&local[]=2",
+            true,
+            false
+          );
+          cliExecute("set _voteToday = true");
+        } else cliExecute("VotingBooth.ash");
+      },
```
Comment (Pantocyclus) [inline src/tasks/runstart.ts:383]: I don't like the idea of this. There was a considering way back where I would like it to be possible to run this as a 2-(or more) Day script, and this hardcoding here could be potentially very bad if we picked the bad choices. As mentioned before, we should change this into a structure similar to garbo

File: src/tasks/runstart.ts (hunk 17 above)
Comment (Pantocyclus) [reply src/tasks/runstart.ts:149]: Also, if we aren't defaulting to always picking codpiece, the default outfit should probably be updated to not use the codpiece by default if we aren't a myst class

File: src/tasks/leveling.ts [hunk 21]
```diff
@@ -754,33 +878,37 @@ export const LevelingQuest: Quest = {
...
+          `Completed condition: ${!have(reagentBoosterIngredient) && !have(reagentBoosterItem)}`
+        );
+        if (!have(reagentBoosterItem)) {
           if (get("reagentSummons") === 0) useSkill($skill`Advanced Saucecrafting`, 1);
-          create($item`ointment of the occult`, 1);
+          create(reagentBoosterItem, 1);
         }
-        ensureEffect($effect`Mystically Oiled`);
+        ensureEffect(reagentBoosterEffect);
       },
```
Comment (Pantocyclus) [inline src/tasks/leveling.ts:894]: As a non-sauceror, we are only getting a single potion from crafting. Is this sufficient for a low shiny non-sauceror?

File: src/tasks/leveling.ts [hunk 22]
```diff
@@ -754,33 +878,37 @@ export const LevelingQuest: Quest = {
...
-          use($item`oil of expertise`, itemAmount($item`oil of expertise`) - 1);
-        if (have($item`cherry`) && have($effect`Expert Oiliness`))
-          putCloset(itemAmount($item`cherry`), $item`cherry`);
+        if (itemAmount(reagentBalancerItem) > 1)
+          use(reagentBalancerItem, itemAmount(reagentBalancerItem) - 1);
+        if (have(reagentBalancerIngredient) && have(reagentBalancerEffect))
+          putCloset(itemAmount(reagentBalancerIngredient), reagentBalancerIngredient);
       },
       limit: { tries: 1 },
     },
```
[the remark this answers is not in the document]
Comment (Ignose) [reply src/tasks/leveling.ts:949]: We can't. I haven't found a way to fix this yet.

File: src/tasks/leveling.ts (hunk 21 above)
Comment (Ignose) [reply src/tasks/leveling.ts:929]: As a non-sauceror, this will not be sufficient if the runner has insufficient sources of other mainstat boost. It will not run as a non-Sauceror with the minimum requirements.
